_posts/2025-09-05-anatomy-of-vllm.md

```diff
@@ -0,0 +1,992 @@
+
+```python
+from vllm import LLM, SamplingParams
+
+prompts = [
+    "Hello, my name is",
+    "The president of the United States is",
+]
+
+sampling_params = SamplingParams(temperature=0.8, top_p=0.95)
+
+def main():
+    llm = LLM(model="TinyLlama/TinyLlama-1.1B-Chat-v1.0")
+
+    outputs = llm.generate(prompts, sampling_params)
+
+if __name__ == "__main__":
+    main()
+```
+
+> [!NOTE]
+> Environment vars:
+> * VLLM_USE_V1="1" # we're using engine V1
+> * VLLM_ENABLE_V1_MULTIPROCESSING="0" # we're running in a single process
+
+This configuration is:
+
+* offline (no web/distributed system scaffolding)
+* synchronous (all execution happens in a single blocking process)
+* single-GPU (no data/model/pipeline/expert parallelism; DP/TP/PP/EP = 1)
+* using standard transformer [[2]](#ref-2) (supporting hybrid models like Jamba requires a more complex hybrid KV-cache memory allocator)
+
+From here, we'll gradually build up to an online, async, multi-GPU, multi-node inference system - but still serving a standard transformer.
+
+In this example we do two things, we:
+
+1. Instantiate an engine
+3. Call <code>generate</code> on it to sample from the given prompts
+
+Let's start analyzing the constructor.
+
+### LLM Engine constructor
+The main components of the engine are:
+
+* vLLM config (contains all of the knobs for configuring model, cache, parallelism, etc.)
+* processor (turns raw inputs → <code>EngineCoreRequests</code> via validation, tokenization, and processing)
+* engine core client (in our running example we're using <code>InprocClient</code> which is basically == <code>EngineCore</code>; we'll gradually build up to <code>DPLBAsyncMPClient</code> which allows serving at scale)
+* output processor (converts raw <code>EngineCoreOutputs</code> → <code>RequestOutput</code> that the user sees)
+> [!NOTE]
+> With the V0 engine being deprecated, class names and details may shift. I'll emphasize the core ideas rather than exact signatures. I'll abstract away some but not all of those details.
+
+Engine core itself is made up of several sub components:
+
+* Model Executor (drives forward passes on the model, we're currently dealing with <code>UniProcExecutor</code> which has a single <code>Worker</code> process on a single GPU). We'll gradually build up to <code>MultiProcExecutor</code> which supports multiple GPUs
+* Structured Output Manager (used for guided decoding - we'll cover this later)
+* Scheduler (decides which requests go into the next engine step) - it further contains:
+  <ol type="a">
+  <li>policy setting - it can be either <b>FCFS</b> (first come first served) or <b>priority</b> (higher priority requests are served first)</li>
+  <li><code>waiting</code> and <code>running</code> queues</li>
+  <li>KV cache manager - the heart of paged attention [[3]](#ref-3)</li>
```

Comment:
this reference rendering seems to be broken inside `<li></li>`

Reply:
fixed! will push out soon